lib/improver/nowcasting/accumulation.py

```diff
@@ -0,0 +1,270 @@
+accumulations from advected radar fields. It also includes the
+AccumulationAggregator class for combining these into longer periods.
+"""
+import warnings
+import numpy as np
+
+import iris
+from improver.cube_combiner import CubeCombiner
+from improver.utilities.cube_units import (enforce_coordinate_units_and_dtypes,
+                                           enforce_diagnostic_units_and_dtypes)
+
+
+class Accumulation:
+    """
+    Class to calculate precipitation accumulations from radar rates fields
+    provided at discrete time intervals. The plugin will return one fewer
+    cubes of accumulation than the number of input precipitation rate cubes.
+    The time intervals between the precipitation rate cubes determine the
+    accumulation periods.
+    """
+    def __init__(self, accumulation_units='m'):
+        """
+        Initialise the plugin.
+
+        Args:
+            accumulation_units (str):
+                The physical units in which the accumulation should be
+                returned. The default is metres.
+        """
+        self.accumulation_units = accumulation_units
+
+    def __repr__(self):
+        """Represent the plugin instance as a string."""
+        result = '<Accumulation: accumulation_units={}>'
+        return result.format(self.accumulation_units)
+
+    @staticmethod
+    def sort_cubes_by_time(cubes):
+        """
+        Sort cubes in time ascending order (from earliest time to latest time).
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing input precipitation rate cubes.
+        Returns:
+            cubes (iris.cube.CubeList):
+                The cubelist in ascending time order.
+            times (list):
+                A list of the validity times of the precipitation rate cubes in
+                integer seconds since 1970-01-01 00:00:00.
+        """
+        times = np.array([cube.coord('time').points[0] for cube in cubes])
+        time_sorted = np.argsort(times)
+        times = times[time_sorted]
+        cubes = iris.cube.CubeList(np.array(cubes)[time_sorted])
+        return cubes, times
+
+    def process(self, cubes):
+        """
+        Calculate period precipitation accumulations based upon precipitation
+        rate fields.
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing input precipitation rate cubes.
+        Returns:
+            accumulation_cubes (iris.cube.CubeList):
+                A cubelist containing precipitation accumulation cubes where
+                the accumulation periods are determined by the intervals
+                between the precipitation rate cubes.
+        """
+        # Standardise inputs to expected units.
+        cubes = enforce_coordinate_units_and_dtypes(
+            cubes, ['time', 'forecast_reference_time', 'forecast_period'],
+            inplace=False)
+        enforce_diagnostic_units_and_dtypes(cubes)
```

Comment:
I'm undecided as to whether this design is "safe" or not.  I can see the principle - we "expect" certain datatypes and our plugins assume that.  But I'm worried about the practise, and the downstream impacts of any changes to expected units etc.

If we're taking this approach, I think for the moment the expected units should be documented (in comments or docstrings), so that a developer looking at this can see that yes, rates in `m s-1` multiplied by time in `s` makes accumulations in `m`.

Reply:
In this instance the changes are applied to copies of the cubes, so the originals are unaffected. As to the output, the final units are specified by a keyword argument (accumulation_units), so the user has control over what is returned from this plugin. By defaulting to our expected working units the code can be that much simpler and the conversion to the users desired outputs can be a single line step at the end of the process.